design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+--   the same grammar question as above (ie, CREATE MV statement) raises (again, not important to the core design we want to discuss here)
+```
+
+Maybe in the future:
+
+```sql
+CREATE TABLE rdbmsStyleTable <withSomeSchemaSpec>
+-- create an empty persistent TABLE that is only modifies via insert/update/delete statements
+-- currently not supported by KSQL
+```
+
+```sql
+CREATE TABLE rdbmsStyleTable AS SELECT * FROM someTableOrView -- note, it does not make a difference if the input on the right hand side SELECT-query is a "table" or a MV
+-- create a table with initial state being the result of the query (same as a RDBMS query)
+```
+
+
+**Example 100: a stateless stream query**
+
+```sql
+[CREATE STREAM resultStream AS] SELECT * FROM inputStream
+-- returns a STREAM
+-- a client can issue a transient query to get an infinite result stream it can iterator over (until the client terminates the query) or
+-- the result stream can be persistent into an output STREAM (that is backed by a topic)
+-- same as current KSQL syntax (we can add filters, projections, etc)
+-- the query starts to consume inputStream for "beginning" and does not terminate
+```
+
+The last point is a design principle in our language:
+
+* (P1) a new query over an input stream by default starts to query the input stream from the beginning
+* (P1-a) a user can set a different "start point" with a filter predicate, e.g., `ROWTIME > NOW()` (more details later)
+* (P1-b) a user can define a "termination point" with a filter predicate, e.g., `ROWTIME < NOW() + 1 DAY` (more details later)
+  * It's open for discussion, if a "termination query over an input stream" can be a persistent query or not
+    (ie, should we only support transient terminal queries?)
+  * My personal take is: we can allow both types, because we already have `INSERT INTO myStream` and we can generalize it –
+    a STREAM may or may not have a continuous query that write into it – at the same time, a STREAM may have multiple CQ that write into it,
+    for which the writes are non-deterministically interleaved – IMHO, it's even "ok" to write into a STREAM from "external"
+    by writing into the corresponding topic (it's risky though to not break the schema...) –
+    of course, users need to be aware of what they are doing, but I advocate for maximum flexibility
+
+
+**Example 150: stream-table join**
+
+```sql
+[CREATE STREAM resultStream AS] SELECT * FROM inputStream s JOIN inputTable t ON s.key = t.key
+-- same properties as for Example 100
+-- because one input is a stream, the output is a stream
+```
+
+The last point is a design principle in our language:
+
+* (P2) a query with at least one input STREAM, returns a result STREAM
+* (P2-a) we can extend this principle to (stream-table)-table, stream-(table-table) and (stream-stream)-table joins (and any other n-ary query)
+
+The reason for this design principle is, that the "infinite" nature of an input stream dominates the "finite" nature of a table –
```

Comment:
Aren't both tables and streams infinite? A table gets continuously updated. Isn't their difference the deduplication/upserts? Whether a join returns a stream or a table should be a knob: do you want to see new records in the output or do you want to see updates if the same key already exists

Reply:
> Aren't both tables and streams infinite?

Not from my point of view. The state of a TABLE is a finite set of rows. And a query is _always_ against the table state. If you want to query the changelog, you need to use the `STREAM(...)` operator.

> Whether a join returns a stream or a table should be a knob: do you want to see new records in the output or do you want to see updates if the same key already exists

I don't think that you can have a knob what the result type is; the result type is defined by the operator. And Stream-Stream, Stream-Table, and Table-Table joins are quite different. What I propose is as follows:

Table-Table join: behaves like an RDBMS query; the (finite) state of both tables are joined resulting is a final result (the query terminates; updates to the base tables are irrelevant; the result is _not_ a materialized view that get's updates if the base data changes; ie, the query has "snapshot" semantics):

Stream-Table join: this is a lookup join; for each stream record a table-lookup is performed to enrich the stream record. Hence, the output is data stream.

Stream-Stream join: _alyways_ windowed; returns an output stream

Reply:
> Table-Table join: behaves like an RDBMS query; the (finite) state of both tables are joined resulting is a final result (the query terminates; updates to the base tables are irrelevant; the result is not a materialized view that get's updates if the base data changes; ie, the query has "snapshot" semantics)

Confused, classic db can do table-table join to create a MV.  Are you saying this wouldn't be supported? That doesn't fit with how I read your doc!

Reply:
Create a MV view for a table-table-join would be supported, but this would be _persistent_ query, while my comment was about _transient_ queries.

Hence: `SELECT * FROM table1 JOIN table2 ON t1.key = t2.key` would be transient query (ie, pull query). It computes a result based on the table state of both input table (we need to define _exacty_ semantics which version of the state we use) and terminates.

To create a MV, you would create a `stream()` on the result table and upsert it into the result MV:
```
// full clumsy syntax that we could simplify with implicit conversions
CREATE MV view AS SELECT * FROM STREAM(SELECT * FROM table1 JOIN table2 ON t1.key = t2.key)
// maybe (allow `STREAM(<table>)` to be a full query)
CREATE MV view AS STREAM(SELECT * FROM table1 JOIN table2 ON t1.key = t2.key)
// or even (implicit conversion from the query result table into a stream that is upserted)
CREATE MV view AS SELECT * FROM table1 JOIN table2 ON t1.key = t2.key
```